### ISSUE TITLE
Pinning is slow with s3 datastore

### ISSUE THREAD

**Issue by godcong** [ipfs/go-ipfs#9018]
### Checklist

- [X] My issue is specific & actionable.
- [X] I am not suggesting a protocol enhancement.
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my issue.

### Description

I manage my ipfs data through S3, and multiple servers share a common S3
But my index list is stored locally.
When I lose my index list, I can only reopen pin all the data.
I would like to have an option to set at pin time to check if the data exists before pin.

**Comment by Jorropo** [comment]
@godcong I'm not sure to understand your question.

What does:
> set at pin time to check if the data exists before pin

Means ?
What does it check ? (that the data is available in the datastore ? Available in the network ?)
What happen if that check fails ?
What do you hope gaining for it ?

Checking if the data exists require traversing the dag, checking before pinning would require a double traversal would be very slow.

Pin time and checking time or in fact the same step and we don't want to seperate it. (and pinning is transactional so if an issue is happening we can just cancel it).

**Comment by godcong** [comment]
@Jorropo 
I use s3 to manage `block`. When a file pin is completed, these `block` are already stored in s3.
I save only `datastore` data locally.
But I found that when my server went down and I reconfigured ipfs.
I just lost the data in the datastore, the block is still intact.
But the data that had been pin completed was at the same rate as the first pin.
So, is it necessary to check if the block is complete and saved before ipfs saves the block?
